developing/cicd/README.md

```diff
@@ -8,9 +8,175 @@ actually be used by anyone. This section provides resources for setting
+
+### Build
+
+**Builds should be repeatable.**
+This means you should be able to check out the code from your project at that same commit hash and build it again and get the same artifact(s).
+_Note_: This means that dependencies should be versioned in your codebase.
+To keep dependencies up to date in an automated way you can try [Dependabot](https://dependabot.com/).
+Our internal documentation on setting up Dependabot is [here](./dependabot.md).
+Please do whatever is "correct" in the context of the languages/frameworks your project is built in.
+
+**Builds should be hermetic.** This means that the build should be isolated from other builds.
+In CI, the build shouldn't share the same workspace or files from a previous build or from a build of a different project.
+
+**Builds should output an immutable artifact.** Artifact outputs should not be rewritten or altered by subsequent builds.
+This allows you to distribute or redeploy from these unaltered artifacts for different points in code.
+Additionally, try to label the artifact with the appropriate versioning scheme for your project.
+
+#### Notes about Versioning
+
+Version your code and artifacts. Doesn't really matter how, just do it. It will make it easier to track down issues or deploy specific versions of your project.
+
+In your code consider adding a command or an endpoint that surfaces that version for debugging.
+
+Tag your versioned releases on your mainline branch.
+This helps you find the state of mainline at the point in time your artifact was built.
+Additionally, if you are using GitHub you can use their releases functionality to share release artifacts and notes.
+
+Here is a preferential ordering of versioning schemes:
+
+##### Semantic Versioning
+
+Why semver? It tells you and your customer how much has changed since the last released code and sets expectations accordingly.
+If you are tagging at mainline where you are building from, you can rebuild the artifact from the same point.
+
+From [semver.org](https://semver.org/).
+
+```
+Given a version number MAJOR.MINOR.PATCH, increment the:
+
+MAJOR version when you make incompatible API changes,
+MINOR version when you add functionality in a backwards compatible manner, and
+PATCH version when you make backwards compatible bug fixes.
+Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.
+```
+
+##### Commit Hash
+
+A commit hash is unique (with extremely few collisions) and is easily linked back to history in code.
+However, how much has changed is opaque to your users and it is difficult to determine how old this version is in comparison to other versions.
+
+##### Other versioning strategies
+
+These are other versioning strategies we've seen. We do not recommend them.
+
+* _Feature branch related names._ These should be short lived and not maintained for more than a week at a time. Most users will not find these useful.
```

Comment:
Curious how you came up with "a week".

Reply:
It was kinda arbitrary. I think I read something like that timeline in accelerate.... that was a long time ago.

Reply:
I'd maybe push for "maintained for no more than a few days." A whole week sounds like an outlier to me.